plan/gen_physical_plans.go

```diff
@@ -271,32 +280,116 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
-		return p.constructIndexJoin(prop, innerJoinKeys, outerJoinKeys, outerIdx, innerPlan)
+	if bestIndexInfo != nil {
+		innerPlan := x.forceToIndexScan(bestIndexInfo, remainedOfBest)
+		return p.constructIndexJoin(prop, orderedInnerKeys, orderedOuterKeys, outerIdx, innerPlan, rangesOfBest, idxInIndex)
 	}
 	return nil
 }
 
+// checkIndexForIndexJoin checks whether this index can be used for building index join and return enough information to
+// determine whether this index is better and how to construct the index join.
+func (p *LogicalJoin) checkIndexForIndexJoin(indexInfo *model.IndexInfo, innerPlan *DataSource, innerJoinKeys []*expression.Column) (ok bool,
+	remained []expression.Expression, indexRanges []*ranger.IndexRange, idxOff2KeyOff []int) {
+	idxCols, colLengths := expression.IndexInfo2Cols(innerPlan.Schema().Columns, indexInfo)
+	if len(idxCols) == 0 {
+		return false, nil, nil, nil
+	}
+
+	idxOff2KeyOff = joinKeysMatchIndex(innerJoinKeys, indexInfo)
+	if idxOff2KeyOff == nil {
+		return false, nil, nil, nil
+	}
+	maxIndexColIdx := 0
+	for i := len(indexInfo.Columns) - 1; i >= 0; i-- {
+		if idxOff2KeyOff[i] != -1 {
+			maxIndexColIdx = i
+			break
+		}
+	}
+
+	// After predicate push down, the one side conditions of join must be the conditions that cannot be pushed down so cannot
+	// to calculate range either. So we only need the innerPlan.pushedDownConds and the eq conditions that we generate.
+	// TODO: There may be a selection that block the index join.
+	conds := make([]expression.Expression, 0, len(idxOff2KeyOff)+len(innerPlan.pushedDownConds))
+	// Construct a fake equal expression for calculating the range.
+	for _, key := range innerJoinKeys {
+		// Int datum 1 can convert to all column's type(numeric type, string type, json, time type, enum, set) safely.
+		fakeConstant := &expression.Constant{Value: types.NewIntDatum(1), RetType: key.GetType()}
+		eqFunc := expression.NewFunctionInternal(p.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), key, fakeConstant)
+		conds = append(conds, eqFunc)
+	}
+
+	lastExprHashCode := conds[idxOff2KeyOff[maxIndexColIdx]].HashCode()
+
+	conds = append(conds, innerPlan.pushedDownConds...)
+	// After constant propagation, there won'be cases that t1.a=t2.a and t2.a=1 occur in the same time.
+	// And if there're cases like t1.a=t2.a and t1.a > 1, we can also guarantee that t1.a > 1 won't be chosen as access condition.
+	// So DetachIndexConditions won't miss the equal conditions we generate.
+	accesses, remained := ranger.DetachIndexConditions(conds, idxCols, colLengths)
+
+	// We should guarantee that all the join's equal condition is used. Check that last one is in the access conditions is enough.
+	// Here the last means that the corresponding index column's position is maximum.
+	valid := false
+	for _, expr := range accesses {
+		if bytes.Equal(expr.HashCode(), lastExprHashCode) {
+			valid = true
+			break
+		}
+	}
+	if !valid {
+		return false, nil, nil, nil
+	}
+
```

Comment:
From line 351 to 382, there is a black magic. Please extract a function and add enough comments.